geode-core/src/distributedTest/java/org/apache/geode/distributed/RequestDistributedLockWhileClosingCacheDistributedTest.java

```diff
@@ -0,0 +1,184 @@
+    distributionManager.addMembershipListener(observer);
+  }
+
+  private void disconnectDistributedSystem() throws InterruptedException, TimeoutException {
+    // Wait for the grantor to signal ABOUT_TO_PROCESS_LOCK_REQUEST
+    blackboard.waitForGate(ABOUT_TO_PROCESS_LOCK_REQUEST);
+
+    // Disconnect the distributed system
+    ClusterStartupRule.getCache().getInternalDistributedSystem().disconnect();
+  }
+
+  private void requestDistributedLock(String lockName) {
+    // Request a distributed lock from the partitioned region lock service
+    DistributedLockService service =
+        ClusterStartupRule.getCache().getPartitionedRegionLockService();
+    try {
+      service.lock(lockName, GeodeAwaitility.getTimeout().toMillis(), -1);
+    } catch (Exception e) {
+      /* ignore */
+    }
+  }
+
+  private void verifyCacheIsClosed() {
+    // Verify the cache is closed
+    assertThat(ClusterStartupRule.getCache().isClosed()).isTrue();
+  }
+
+  private void verifyLockServiceDoesNotHoldToken(Object lockName) {
+    // Verify no the lock token is not held by the grantor server after the lock requesting server
+    // has departed
+    DLockService dLockService =
+        (DLockService) ClusterStartupRule.getCache().getPartitionedRegionLockService();
+    assertThat(dLockService.isLockGrantor()).isTrue();
+    DLockGrantor dLockGrantor = dLockService.getGrantor();
+    DLockGrantor.DLockGrantToken grantToken = dLockGrantor.getGrantToken(lockName);
+    // assertThat(grantToken).isNull();
+  }
+
+  class TestDistributionMessageObserver extends DistributionMessageObserver implements
+      MembershipListener {
+
+    private final String lockName;
+
+    public TestDistributionMessageObserver(String lockName) {
+      this.lockName = lockName;
+    }
+
+    public void memberDeparted(DistributionManager distributionManager,
+        InternalDistributedMember id, boolean crashed) {
+      // Signal the member has departed. This will cause the DLockRequestMessage to be processed.
+      blackboard.signalGate(MEMBER_DEPARTED);
+    }
+
+    public void beforeProcessMessage(ClusterDistributionManager dm, DistributionMessage message) {
+      if (message instanceof DLockRequestProcessor.DLockRequestMessage) {
+        DLockRequestProcessor.DLockRequestMessage dLockRequestMessage =
+            (DLockRequestProcessor.DLockRequestMessage) message;
+        if (dLockRequestMessage.getObjectName().equals(this.lockName)) {
+          // Signal the about to process lock request. This will cause the lock requesting server to
+          // disconnect its distributed system.
+          blackboard.signalGate(ABOUT_TO_PROCESS_LOCK_REQUEST);
+
+          // Wait for member departed before processing the DLockRequestMessage
+          try {
+            blackboard.waitForGate(MEMBER_DEPARTED);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
```

Comment:
Please use `DistributedErrorCollector` for any exceptions in an asynchronous context like this instead of just rethrowing it. It'll collect up all exceptions from all DUnit VMs and report them when the test finishes (including causing it to "FAIL" instead of "PASS"). Rethrowing might or probably results in a failure depending on the surrounding product call that invokes this code but the error collector is a lot cleaner. 

Non-dunit tests can use JUnit's `ErrorCollector`.

Reply:
Thanks a lot, Kirk. I didn't know we had this rule. I just add the exception to the DistributedErrorCollector like this, right?
```
} catch (Exception e) {
  errorCollector.addError(e);
}
```

Reply:
Yep! And if you need to you can always rethrow the exception after adding to errorCollector so that GemFire reacts to it:
```
} catch (Exception e) {
  errorCollector.addError(e);
  throw e;
}
```